ipld/unixfs/io/directory.go

```diff
@@ -608,14 +821,23 @@ func (d *DynamicDirectory) RemoveChild(ctx context.Context, name string) error {
 		return hamtDir.RemoveChild(ctx, name)
 	}
 
-	basicDir, err := hamtDir.switchToBasic(ctx)
+	// We have not removed the element that violates MaxLinks, so we have to +1 the limit. We -1 below.
+	basicDir, err := hamtDir.switchToBasic(ctx, WithMaxLinks(hamtDir.maxLinks+1), WithCidBuilder(hamtDir.GetCidBuilder()))
 	if err != nil {
 		return err
 	}
 	err = basicDir.RemoveChild(ctx, name)
 	if err != nil {
 		return err
 	}
-	d.Directory = basicDir
+
+	basicDir.setMaxLinks(hamtDir.maxLinks - 1)
+	d.directoryWithOptions = basicDir
 	return nil
 }
+
+// validShardWidth verifies that the given number is positive, a power of 2
+// and a multiple of 8.
+func validShardWidth(n int) bool {
+	return n > 0 && (n&(n-1)) == 0 && n%8 == 0
```

Comment:
Since doing the bitwise check for mod 2 (`n&(n-1)`) shows that the number is a power of 2, it is safe to calculate the bitwise modulus: n%m --> n&(m-1)
```suggestion
	return n > 0 && (n&(n-1)) == 0 && n&7 == 0
```